### DISCUSSION TITLE
Add ability for DeepSleepTime to start at midnight of the current day

### DISCUSSION THREAD

**Discussion by jsiddall** [arendst/Tasmota#17813] [Ideas]
The current implementation of `DeelSleepTime` wakes at a specific period, based on the epoch (midnight Jan 1 1970 in the Tasmota timezone).  This works fine for `DeepSleepTime` that are even multiples of hours or days but doesn't easily allow choosing a specific wake time (ex 10:00) today.

What about creating a new SetOption that, when enabled, starts the sleep time calculations at midnight of the current day instead of the epoch.  The default value of the option would be 0 which would preserve the current behavior and thus be backwards compatible.

Enabling the option would allow a setting like `DeepSleepTime 36000` which would wake at 10:00 and 20:00.  If the calculation went past midnight (i.e. next wake time more than 86400 seconds since midnight) then reset the calculation to start at midnight tomorrow.

Discussion of this idea started in docs PR [#1158](https://github.com/tasmota/docs/pull/1158)

**Comment by stefanbode** [comment]
Yes there are a a required wake-up every 60min. But beside a small power change and a 1 second up you should not see anything. At least there should be no wlan connect or any mqtt connect. The small wake-up is mandatory because of some limitations. See the documentation. If we allow 3 or more days we have this 1970 discussion most people do not understand. As mentioned a "deepsleeptime 86400 14:45" to wake-up daily at 2:45pm is something that may work easily. The variable is 32bit. If we restrict the deepsleep to 16bit 64.000 we can use the other 16bit for a shift of deepsleep without requesting an additional 16bytes in the permanent storage.

**Reply by jsiddall** [reply]
Agree, the Tasmota does wake frequently (not sure if it is every 71 minutes or 3 hours and something, but it definitely wakes many times a day) to reset the sleep time.  I can confirm it doesn't seem to join WLAN at these "mini wakes".

I think you start having to understand the epoch thing any time you get over 1 day (86400 s) because the first sleep might not be until midnight two days from now instead of tomorrow midnight -- which most people would expect.

But you could also run into the "epoch problem" anytime you get non-even hour divisions of 24.  Ex: a `DeepSleepTime` of 18000 seconds will wake at different times today than it will tomorrow.

Regarding splitting the 32 bit int into two 16 bits, the problems I can see with that are:

1. With 16 bits you can't set values as large as one day (86400 seconds per day vs 65535 possible values in a 16 bit uint)
2. It would break backward compatibility with any existing Tasmotas using `DeepSleepTime` greater than about 18 hours, of which I am sure there are many

**Reply by stefanbode** [reply]
Agree. 16bit is not enough and this was also the reason using 32 bits instead of 16 bit. We have 24x60=1440 minutes a day 8192 is 13bit. So we have 19bit for the deepsleep 524.000 seconds. Should be enough

**Reply by barbudor** [reply]
I do agree that wake-up at a given time is a valid use case.
Ideally we could have `DeepSleepPeriod` for the current behavior and `DeepSleepTime` for that but that would be a breaking change
It could be `DeepSleepTime N` as today for the period and `DeepSleepTime HH:MM` for the 'at a given time'
Could be easily detected by the presence of `:`

But the big thing is how ot modify the algorithm.
I'm not very tempted to put a finger in that as I undertstand the regulation code is tricky to adjust the proper sleep duration afetr a few wake ups
@stefanbode I thing you already worked on that ?

**Reply by jsiddall** [reply]
I like the idea of using a different syntax with the colon for the TOD vs. the integer for the interval.  Easier than having to set an option while still keeping backward compatibility.

I agree, a nice alternative would have been to use `DeepSleepPeriod` for what we have today and then use `DeepSleepTime` for the ToD version, but unfortunately hindsight doesn't solve this!

**Reply by stefanbode** [reply]
Yes, i did the math before and feel fine with minimum impact to change this. Currently on holiday. Maybe find some time to do it.

**Reply by pkkrusty** [reply]
Did this ever happen? Haven't noticed anything in PRs, but I fully support the expansion to deepsleep. Very useful for me.

**Reply by stefanbode** [reply]
no, not yet. the ESP32 shutter get all my attention :-). I supposed to add this to the scheduler we already have but this makes more headache because the ON/OFF/TOGGLE/and something else already occupy the 2bit Array for the status.
